server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java

```diff
@@ -0,0 +1,250 @@
+    public CCSUsageTelemetry() {
+        this.byRemoteCluster = new ConcurrentHashMap<>();
+        totalCount = new LongAdder();
+        successCount = new LongAdder();
+        failureReasons = new ConcurrentHashMap<>();
+        took = new LongMetric();
+        tookMrtTrue = new LongMetric();
+        tookMrtFalse = new LongMetric();
+        remotesPerSearch = new LongMetric();
+        skippedRemotes = new LongAdder();
+        featureCounts = new ConcurrentHashMap<>();
+        clientCounts = new ConcurrentHashMap<>();
+    }
+
+    public void updateUsage(CCSUsage ccsUsage) {
+        assert ccsUsage.getRemotesCount() > 0 : "Expected at least one remote cluster in CCSUsage";
+        // TODO: fork this to a background thread?
+        doUpdate(ccsUsage);
+    }
+
+    // This is not synchronized, instead we ensure that every metric in the class is thread-safe.
+    private void doUpdate(CCSUsage ccsUsage) {
+        totalCount.increment();
+        long searchTook = ccsUsage.getTook();
+        if (isSuccess(ccsUsage)) {
+            successCount.increment();
+            took.record(searchTook);
+            if (isMRT(ccsUsage)) {
+                tookMrtTrue.record(searchTook);
+            } else {
+                tookMrtFalse.record(searchTook);
+            }
+            ccsUsage.getPerClusterUsage().forEach((r, u) -> byRemoteCluster.computeIfAbsent(r, PerClusterCCSTelemetry::new).update(u));
+        } else {
+            failureReasons.computeIfAbsent(ccsUsage.getStatus(), k -> new LongAdder()).increment();
+        }
+
+        remotesPerSearch.record(ccsUsage.getRemotesCount());
+        if (ccsUsage.getSkippedRemotes().isEmpty() == false) {
+            skippedRemotes.increment();
+            ccsUsage.getSkippedRemotes().forEach(remote -> byRemoteCluster.computeIfAbsent(remote, PerClusterCCSTelemetry::new).skipped());
+        }
+        ccsUsage.getFeatures().forEach(f -> featureCounts.computeIfAbsent(f, k -> new LongAdder()).increment());
+        String client = ccsUsage.getClient();
+        if (client != null && KNOWN_CLIENTS.contains(client)) {
+            // We count only known clients for now
+            clientCounts.computeIfAbsent(ccsUsage.getClient(), k -> new LongAdder()).increment();
+        }
+    }
+
+    private boolean isMRT(CCSUsage ccsUsage) {
+        return ccsUsage.getFeatures().contains(MRT_FEATURE);
+    }
+
+    private boolean isSuccess(CCSUsage ccsUsage) {
+        return ccsUsage.getStatus() == Result.SUCCESS;
+    }
+
+    public Map<String, PerClusterCCSTelemetry> getTelemetryByCluster() {
+        return byRemoteCluster;
+    }
+
+    /**
+     * Telemetry of each remote involved in cross cluster searches
+     */
+    public static class PerClusterCCSTelemetry {
+        private final String clusterAlias;
+        // Right now, this is the number of successful (not skipped) requests to this cluster.
+        // We need to make it clear in the docs that it does not count skipped requests.
```

Comment:
The second sentence of this comment should probably become a JIRA ticket so we can remove it from the code before merging to main.